tacs/mphys/solver.py

```diff
@@ -0,0 +1,196 @@
+                update = True
+
+        tmp = update
+        # Perform all reduce to check if any other procs came back True
+        update = self.comm.allreduce(tmp)
+        return update
+
+    def _update_internal(self, inputs, outputs=None):
+        if self._need_update(inputs):
+            self.sp.setDesignVars(inputs["tacs_dvs"])
+            self.sp.setNodes(inputs["x_struct0"])
+        if outputs is not None:
+            self.sp.setVariables(outputs[self.states_name])
+        self.sp._updateAssemblerVars()
+
+    def apply_nonlinear(self, inputs, outputs, residuals):
+        self._update_internal(inputs, outputs)
+
+        if self.coupled:
+            Fext = inputs[self.rhs_name]
+        else:
+            Fext = None
+
+        self.sp.getResidual(res=residuals[self.states_name], Fext=Fext)
+
+    def solve_nonlinear(self, inputs, outputs):
+        self._update_internal(inputs)
+
+        if self.coupled:
+            Fext = inputs[self.rhs_name]
+        else:
+            Fext = None
+
+        self.sp.solve(Fext=Fext)
+        self.sp.getVariables(states=outputs[self.states_name])
+
+    def solve_linear(self, d_outputs, d_residuals, mode):
+        if mode == "fwd":
+            if self.check_partials:
+                print("solver fwd")
+            else:
+                raise ValueError("forward mode requested but not implemented")
+
+        if mode == "rev":
+            self.sp.solveAdjoint(
+                d_outputs[self.states_name], d_residuals[self.states_name]
+            )
+
+    def apply_linear(self, inputs, outputs, d_inputs, d_outputs, d_residuals, mode):
+        self._update_internal(inputs, outputs)
+        if mode == "fwd":
+            if not self.check_partials:
+                raise ValueError("TACS forward mode requested but not implemented")
+
+        if mode == "rev":
+            if self.states_name in d_residuals:
+                if self.states_name in d_outputs:
+                    self.sp.addTransposeJacVecProduct(
+                        d_residuals[self.states_name], d_outputs[self.states_name]
+                    )
+
+                if self.rhs_name in d_inputs:
+                    array_w_bcs = d_residuals[self.states_name].copy()
+                    self.fea_assembler.applyBCsToVec(array_w_bcs)
+                    d_inputs[self.rhs_name] -= array_w_bcs
+
+                if "x_struct0" in d_inputs:
+                    self.sp.addAdjointResXptSensProducts(
+                        [d_residuals[self.states_name]],
+                        [d_inputs["x_struct0"]],
+                        scale=1.0,
+                    )
+
+                if "tacs_dvs" in d_inputs:
+                    self.sp.addAdjointResProducts(
+                        [d_residuals[self.states_name]],
+                        [d_inputs["tacs_dvs"]],
+                        scale=1.0,
+                    )
+
+    def _design_vector_changed(self, x):
```

Comment:
Is this still needed or has it been replaced by `_need_update`?

Reply:
Good point, it never seems to be used. Should be removed now